### PR TITLE
feat(observability-dashboards): ship vLLM inference dashboard JSON (u…

### PR DESCRIPTION
**Pull request by Dericko681**
…nwired)

Import the vLLM 'Comprehensive' Grafana dashboard (uid vllm-perf-dashboard, 22 panels, schemaVersion 38) from the observ working repo into charts/observability-dashboards/files/vllm/, plus the matching dashboards-generator package marker at tools/dashboards/src/dashboards/vllm/__init__.py.

NOT wired: no entry is added to values.yaml's dashboards[] list and no GrafanaDashboard CR is rendered. vLLM is not yet deployed in the ai-helm cluster (ticket #703), so the dashboard would 404 / show empty panels if rendered today. The files exist so the wiring is a one-line values.yaml change once the inference backend lands.

The JSON is the canonical source for now (no Python generator module yet); when one is added it should follow the envoy_ai_gateway pattern (build() -> dict + OUTPUT_PATH) and be registered in tools/dashboards/src/dashboards/main.py so 'uv run dashboards check' catches drift in CI.

## 1. Summary

This PR changes:

- [Change 1]
- [Change 2]
- [Change 3]

It solves:

- [Problem / ticket / story link]

---

## 2. Intent

The intent of this PR is:

> [Explain why this change exists.]

---

### In Scope

- [Included change 1]
- [Included change 2]

### Out of Scope

- [Excluded change 1]
- [Excluded change 2]

---

## 4. Verification

I verified this change by:

- [ ] Running automated tests
- [ ] Running manual tests
- [ ] Checking logs
- [ ] Checking metrics
- [ ] Testing error cases
- [ ] Testing permissions/security behavior
- [ ] Testing rollback or failure behavior, if relevant

Commands run:

```bash
[command 1]
[command 2]
```

Results:

```text
[paste result or link]
```

---

## 5. Screenshots / Evidence

Add evidence here:

* Screenshot: [link]
* Logs: [link]
* Metrics: [link]
* Recording: [link]

---

## 6. Risk Assessment

Risk level:

* [ ] Low
* [ ] Medium
* [ ] High

Potential risks:

* [Risk 1]
* [Risk 2]

Mitigation:

* [Mitigation 1]
* [Mitigation 2]

---

## 7. AI Usage Declaration

AI was used for:

* [ ] Understanding existing code
* [ ] Generating code
* [ ] Refactoring
* [ ] Generating tests
* [ ] Drafting documentation
* [ ] Reviewing the diff
* [ ] Not used

Human verification:

* [ ] I understand every meaningful change in this PR
* [ ] I checked generated code manually
* [ ] I checked generated tests manually
* [ ] I removed unsupported AI assumptions
* [ ] I accept responsibility for this PR

---

## 8. Reviewer Focus

Please focus your review on:

* [ ] Correctness
* [ ] Architecture
* [ ] Security
* [ ] Performance
* [ ] Tests
* [ ] Maintainability
* [ ] Product intent
* [ ] Edge cases

### CODE REVIEW

File: charts/observability-dashboards/values.yaml [hunk 1]
```diff
@@ -156,6 +159,12 @@ dashboards:
     folderRef: collectors
     file: files/collectors/tempo-single-binary.json
     resyncPeriod: 5m
+  # vLLM inference engine dashboard — latency, throughput, engine internals, system health.
+  - name: vllm-dashboard
```
Comment (stephane-segning) [inline charts/observability-dashboards/values.yaml:163]: **P2 — PR description says "NOT wired", but the code IS wired**

The PR body's first paragraph states: *"NOT wired: no entry is added to values.yaml's dashboards[] list and no GrafanaDashboard CR is rendered."* However, this diff adds both the `vllm` folder and this `vllm-dashboard` entry to `dashboards[]` — so the dashboard **is** rendered today.

Either the description is stale (from an earlier draft), or the wiring was added without updating the description. Please update the PR description to match the actual state.

File: tools/dashboards/src/dashboards/main.py [hunk 2]
```diff
@@ -41,6 +41,7 @@
     "dashboards.envoy_ai_gateway.my_usage",
     "dashboards.envoy_ai_gateway.chat_overview",
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/main.py:42]: **P2 — Missing `__init__.py`**

The PR body claims *"the matching dashboards-generator package marker at `tools/dashboards/src/dashboards/vllm/__init__.py`"*, but that file isn't in the diff — the `vllm/` directory only contains `vllm.py`. The sibling `envoy_ai_gateway/` package has an empty `__init__.py`.

Add an empty `tools/dashboards/src/dashboards/vllm/__init__.py` for consistency (and to make the PR description accurate).

File: tools/dashboards/src/dashboards/vllm/vllm.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        .with_panel(_timeseries_panel(title="Prefix Cache Hits/s", unit="short", grid=(8, 12, 12, 39), targets=[_prom_target(f"rate({M_PREFIX_CACHE_HITS}{_SEL}[5m])", legend="{{model_name}}")]))
+        .with_panel(_timeseries_panel(title="Prefix Cache Hit Ratio", unit="percent", grid=(8, 12, 0, 47), thresholds=_thresholds(("green", None), ("yellow", 50)), targets=[_prom_target(f"rate({M_PREFIX_CACHE_HITS}{_SEL}[5m]) / (rate({M_PREFIX_CACHE_HITS}{_SEL}[5m]) + rate({M_PREFIX_CACHE_QUERIES}{_SEL}[5m]))", legend="{{model_name}}")]))
+        .with_panel(_timeseries_panel(title="KV Cache & Preemptions", unit="short", grid=(8, 12, 12, 47), targets=[_prom_target(f"{M_KV_CACHE_USAGE}{_SEL}", legend="Usage %"), _prom_target(f"{M_NUM_PREEMPTIONS}{_SEL}", legend="Preemptions")]))
+        .with_panel(_row("System Health (DCGM exporter not running - GPU metrics unavailable)", y=55))
+    )
+
+
+def build():
+    return json.loads(json.dumps(_dashboard().build(), cls=JSONEncoder))
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/vllm/vllm.py:196]: **P2 — Missing trailing newline (CI will fail)**

The diff shows `\ No newline at end of file` here. The `dashboards-drift.yml` CI runs `uv run ruff format --check .` before the drift check, and ruff format always appends a trailing newline.

This will fail CI immediately. Add a trailing newline.

File: tools/dashboards/src/dashboards/vllm/vllm.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+            "The model_name template variable filters every panel. "
+            "GENERATED — source: tools/dashboards/src/dashboards/vllm/vllm.py."
+        )
+        .timezone("browser")
+        .editable()
+        .tooltip(dm.DashboardCursorSync.CROSSHAIR)
+        .refresh("30s")
+        .time("now-1h", "now")
+        .with_variable(
+            db.QueryVariable("model_name")
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/vllm/vllm.py:159]: **P1 — Duplicate refId "A" in multi-target panels**

This panel (and the "Requests/s" and "KV Cache & Preemptions" panels) define multiple targets that all share `refId: "A"`. Grafana expects unique refIds within a panel — duplicates break query transformations, alerting, and panel export references.

Root cause: `_prom_target()` defaults `ref_id="A"` and these callers don't override it. By contrast, `_percentile_targets()` correctly uses A/B/C/D — the established convention.

Fix: pass explicit `ref_id` values for each target in multi-target panels (e.g. `ref_id="B"` for the "Failed" series here).

File: tools/dashboards/src/dashboards/vllm/vllm.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            .label("Model")
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/vllm/vllm.py:160]: **P1 (continued from previous comment) — same refId issue + wrong "Swapped" metric**

This panel has both problems together:
1. All three targets default to `ref_id="A"` (should be A/B/C).
2. The "Swapped" target reuses `M_NUM_WAITING` instead of the correct `vllm:num_requests_swapped` metric (see comment on line 47).

File: tools/dashboards/src/dashboards/vllm/vllm.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            .datasource(MIMIR_DS)
+            .query(f"label_values({M_REQUEST_SUCCESS}, model_name)")
+            .refresh(dm.VariableRefresh.ON_TIME_RANGE_CHANGED)
+            .sort(dm.VariableSort.ALPHABETICAL_ASC)
+            .multi(True)
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/vllm/vllm.py:165]: **P1 (continued) — duplicate refId "A" in KV Cache & Preemptions**

Both targets here default to `ref_id="A"` (Usage% and Preemptions should be A/B).

File: tools/dashboards/src/dashboards/vllm/vllm.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+M_E2E_LATENCY_SUM = "vllm:e2e_request_latency_seconds_sum"
+M_E2E_LATENCY_COUNT = "vllm:e2e_request_latency_seconds_count"
+M_E2E_LATENCY_BUCKET = "vllm:e2e_request_latency_seconds_bucket"
+M_TTFT_SUM = "vllm:time_to_first_token_seconds_sum"
+M_TTFT_COUNT = "vllm:time_to_first_token_seconds_count"
+M_TTFT_BUCKET = "vllm:time_to_first_token_seconds_bucket"
+M_INTER_TOKEN_BUCKET = "vllm:inter_token_latency_seconds_bucket"
+M_PREFILL_BUCKET = "vllm:request_prefill_time_seconds_bucket"
+M_NUM_RUNNING = "vllm:num_requests_running"
+M_NUM_WAITING = "vllm:num_requests_waiting"
```
Comment (stephane-segning) [inline tools/dashboards/src/dashboards/vllm/vllm.py:47]: **P1 — Wrong metric for "Swapped" queue series**

`M_NUM_WAITING` is defined here, and the "Request Queue" panel (line ~160) uses `M_NUM_WAITING` for *both* the "Waiting" and "Swapped" legends. The result: the "Swapped" line perfectly overlaps "Waiting" — misleading on an operational dashboard.

The correct vLLM metric for swapped requests is `vllm:num_requests_swapped`. Add `M_NUM_SWAPPED = "vllm:num_requests_swapped"` here and use it for the third target in the Request Queue panel.

**Review by stephane-segning — CHANGES_REQUESTED** [review]
## Review Summary

This PR ships the vLLM Grafana performance dashboard (18 data panels + 4 section rows) as both a generated JSON file and a Python generator module, and wires it into the observability chart. The generator faithfully follows the established `envoy_ai_gateway` pattern (OUTPUT_PATH contract, `build() -> dict`, JSONEncoder round-trip, `_RowBuilder` adapter) — good work overall.

**Security: ✅ No new surface.** `$model_name` is a literal Grafana template variable (Python sees `'$model_name'`, not a Python variable). No secrets, no unsafe Python patterns, no user-controlled `OUTPUT_PATH`. The `vllm` folder omits `permissions`, consistent with the `collectors` and `applications` folders — appropriate for engine-level metrics with no per-user data.

**Tests: no regression.** No unit tests exist for any dashboard generator (existing repo pattern); the `dashboards-drift.yml` CI provides the primary guard via `uv run dashboards check` byte comparison. The new module is correctly registered in `main.py`, so drift will be caught — **but CI will currently fail** due to the missing trailing newline (P2-3).

### Blocking issues (request changes)

| Grade | Issue | Fix |
|---|---|---|
| **P1** | "Swapped" queue metric duplicates "Waiting" — `vllm:num_requests_waiting` is used for both. Correct metric is `vllm:num_requests_swapped`. | Add `M_NUM_SWAPPED` constant + use it |
| **P1** | Duplicate `refId: "A"` across multiple targets in 3 panels ("Requests/s", "Request Queue", "KV Cache & Preemptions"). Grafana expects unique refIds. | Pass explicit `ref_id` for each target |
| **P2** | Missing trailing newline in `vllm.py` — `ruff format --check` in CI will fail | Add trailing newline |
| **P2** | Missing `vllm/__init__.py` — PR body claims it exists; sibling package has one | Add empty `__init__.py` |
| **P2** | PR description says "NOT wired" but values.yaml diffs add both folder + dashboard entry — description is stale | Update PR description |

### Nits (mergeable)

- **P3**: PR title says "schemaVersion 38", actual is 39.
- **P3**: "22 panels" counts section headers (18 data + 4 rows).
- **P3**: `vllm:request_success_total{finished_reason="error"}` is semantically odd at the name level — flag for verification when vLLM lands (vLLM may expose a separate `vllm:request_error_total`).

Fix the two P1s and the CI-blocking newline, and this is ready to merge.
